Better error messaging for invalid relative paths on write

I tried to write a generic secret to an invalid path (it was interpreted as a relative path) was not great. I ran into this with the client go api, but here it is illustrated just the same with the cli (distilled down)

```
vault write secret/foo..bar "a=b"
Error writing data to secret/foo...bar: Error making API request.

URL: PUT http://127.0.0.1:8200/v1/secret/foo..bar
Code: 400. Errors:

* internal error
```
The logs contain
```
2017/08/14 18:24:36.162348 [ERROR] core: failed to run existence check: error=existence check failed: key cannot be relative path
```
A command like 
```
vault write secret/foo.bar "a=b"
```
works just fine immediately following

Some issues:
1. The logs say that the "key" cannot be a relative path, but it is in fact the *path* that is a relative path here, not one of the keys from the map. Terminology taken from here: https://www.vaultproject.io/api/secret/generic/index.html#create-update-secret
1. The response is a 400 and yet "internal error" is in the content of the response
1. One could potentially argue that "foo..bar" is not a relative path because the ".." does not come after a "/" For example the file system behaves differently (perhaps better):
```
➜  tmp mkdir x
➜  tmp cd x
➜  x mkdir a..b
➜  x ls
a..b
➜  x
```

--- a/physical/physical_view.go
+++ b/physical/physical_view.go
@@ -1,8 +1,12 @@
 package physical
 
 import (
-	"fmt"
+	"errors"
 	"strings"
+)
+
+var (
+	ErrRelativePath = errors.New("relative paths not supported")
 )
 
 // View represents a prefixed view of a physical backend
@@ -74,7 +78,7 @@ func (v *View) Delete(key string) error {
 // sanityCheck is used to perform a sanity check on a key
 func (v *View) sanityCheck(key string) error {
 	if strings.Contains(key, "..") {
-		return fmt.Errorf("key cannot be relative path")
+		return ErrRelativePath
 	}
 	return nil
 }

--- a/vault/barrier_view.go
+++ b/vault/barrier_view.go
@@ -1,7 +1,7 @@
 package vault
 
 import (
-	"fmt"
+	"errors"
 	"strings"
 
 	"github.com/hashicorp/vault/logical"
@@ -20,6 +20,10 @@ type BarrierView struct {
 	readonly bool
 }
 
+var (
+	ErrRelativePath = errors.New("relative paths not supported")
+)
+
 // NewBarrierView takes an underlying security barrier and returns
 // a view of it that can only operate with the given prefix.
 func NewBarrierView(barrier BarrierStorage, prefix string) *BarrierView {
@@ -32,7 +36,7 @@ func NewBarrierView(barrier BarrierStorage, prefix string) *BarrierView {
 // sanityCheck is used to perform a sanity check on a key
 func (v *BarrierView) sanityCheck(key string) error {
 	if strings.Contains(key, "..") {
-		return fmt.Errorf("key cannot be relative path")
+		return ErrRelativePath
 	}
 	return nil
 }

--- a/vault/request_handling.go
+++ b/vault/request_handling.go
@@ -171,6 +171,9 @@ func (c *Core) handleRequest(req *logical.Request) (retResp *logical.Response, r
 		if errType != nil {
 			retErr = multierror.Append(retErr, errType)
 		}
+		if ctErr == ErrInternalError {
+			return nil, auth, retErr
+		}
 		return logical.ErrorResponse(ctErr.Error()), auth, retErr
 	}
 

